Dlang update

ThomasMader opened the pull request:
###### Motivation for this change

Bring dlang versions to newest versions and refactor everything to be nicer.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

ThomasMader committed:
ldc: 1.12.0 -> 1.14.0

ThomasMader committed:
dmd: 2.084.0 -> 2.084.1

ThomasMader committed:
dtools: 2.084.0 -> 2.084.1

ThomasMader committed:
dub: Disable network tests because they add impurities

ThomasMader commented:
@GrahamcOfBorg build ldc dmd dtools dub

veprbl commented on pkgs/development/tools/build-managers/dub/default.nix:
```diff
@@ -51,9 +51,12 @@ let
+  # They rely on state from servers on the Internet and though
+  # should only be run manually once after changes to the package
+  # were made.
+  dubUnittests = if false then
```
Could we instead disable just those network tests? That would allow to remove fixed-output hack and prevent test results from being cached.

worldofpeace commented on pkgs/development/compilers/dmd/default.nix:
```diff
@@ -1,199 +1,163 @@
+          --set-default CC "${targetPackages.stdenv.cc}/bin/cc"
+
+      cd $out/bin
+      tee dmd.conf << EOF
```
Why not use `writeTextFile` and `lib.generators.toINI`?

ThomasMader replied:
How is that better?

worldofpeace replied:
I'd say it's just a better practice, since it's in nix for this purpose.

veprbl commented on pkgs/development/compilers/ldc/default.nix:
```diff
@@ -18,208 +18,163 @@ let
+        echo "$fn: fixing dylib"
+        install_name_tool -id "$fn" "''${flags[@]}" "$fn"
+      done
     }
```
I see this is a direct copypasta form `pkgs/build-support/setup-hooks/fix-darwin-dylib-names.sh`.

I think what you want can be achieved by moving your test to `installCheckPhase`. If you look at the default order phases are executed
https://github.com/NixOS/nixpkgs/blob/358efb66947504d2195a4f1e25e72265cc7348e3/pkgs/stdenv/generic/setup.sh#L1264-L1267
`installCheckPhase` goes after `fixupPhase`. You can let `fixupPhase` do the `install_name_tool` for you before your test is run in `installCheckPhase`.

ThomasMader replied:
I tried that, it doesn't work as I need the source files for the unit tests and those are not available in this stage anymore.

veprbl replied:
Okay.

veprbl commented on pkgs/development/compilers/dmd/default.nix:
```diff
@@ -1,199 +1,163 @@
+      make -j$NIX_BUILD_CORES -f posix.mak BUILD=release ENABLE_RELEASE=1 PIC=1 INSTALL_DIR=$out DMD=${pathToDmd}
+      cd ../phobos
+      echo ${tzdata}/share/zoneinfo/ > TZDatabaseDirFile
+      echo ${curl.out}/lib/libcurl${curlExtension} > LibcurlPathFile
```
```suggestion
      echo ${curl.out}/lib/libcurl${stdenv.hostPlatform.extensions.sharedLibrary} > LibcurlPathFile
```
Please excuse my ignorance, but don't know why you need ".4" before "dylib". If you really need it, you could use readlink:
```suggestion
      echo `readlink "${curl.out}/lib/libcurl${stdenv.hostPlatform.extensions.sharedLibrary}"` > LibcurlPathFile
```

ThomasMader commented:
Can somebody please check if the current state of this PR works on macOS?
I upgraded my machine to Mojave and dmd does not build anymore.
I would like to know it it was indeed the updgrade to Mojave.

veprbl commented:
dmdBuild fails for me: https://gist.github.com/4e91fbbc81327745b36d81d0bce7946d

veprbl commented:
@GrahamcOfBorg build ldc dmd dtools dub

ThomasMader commented:
Thanks!
Yes that's the problem. What version of macOS do you use?

ThomasMader commented:
Btw. what's going on on hydra? All darwin builds fail with "canceled by user".

veprbl commented:
> Yes that's the problem. What version of macOS do you use?

10.13

veprbl commented:
> Btw. what's going on on hydra? All darwin builds fail with "canceled by user".

All evaluations (even older ones) were stopped just few hour ago I would guess that they were terminated to drop current backlog. I could find mention of this on IRC https://logs.nix.samueldr.com/nixos/2019-02-18#1971702;

ThomasMader commented:
Wow, just wow.

I found the commit which introduced the build failure for dmd: ca1062ea24d7d78ddec59400518fd4bbf7caa685

It's a changed documentation/comment in the make-wrapper. :dizzy_face:

ThomasMader commented:
https://forum.dlang.org/post/diycgerwjablujwsfrcs@forum.dlang.org explains a bit the details about what is dmd trying to do.

veprbl commented:
@ThomasMader This is certainly a nice find, but do you have an explanation for what is going on?

ThomasMader commented:
Not yet but that commit seems not to be the correct one.
When I rebase my PR branch to ca1062e~1 I get the same error.
If I do the same on master branch I get a download for dmd package.
So it's probably something machine specific and I need to redo my bisect because I marked as good when a download was available.

ThomasMader commented:
@GrahamcOfBorg build ldc dmd dtools dub

ThomasMader commented:
Ok let's get this into master for now before the branch off. It's important to update the packages to the new versions and all the other changes make everything much more resilient.

The dmd problem with the "-run" parameter on macOS is not understood yet but I disabled the failing tests for now to get a working build and this situation is better than what is currently in master.
I was able to reproduce that I can not successfully build revisions to which there are packages on hydra from master. So it has something to do with the machine.
I will analyse the problem further and open a PR or issue depending on the outcome and link back to this PR as well.
As I did not include the error message here yet here it is:

```
/var/tmp/dmd_runWK6b1G: Permission denied
core.thread.ThreadError@core/thread.d(3197): Unable to load thread state
```

It happens when dmd is used with the _-run_ parameter which compiles, links and runs the program without leaving any files on the machine. As it is not used as heavily and not needed at all for building other packages it's not a blocker.

ThomasMader commented:
@GrahamcOfBorg build ldc

veprbl commented:
I've tried to rebuild this on macOS. This time I get a failure in dtools: https://gist.github.com/veprbl/3e0be30bd86142418232f6bf1ff57147

veprbl commented on pkgs/development/compilers/dmd/default.nix:
```diff
@@ -1,199 +1,165 @@
-      sha256 = phobosSha256;
-      name = "phobos";
+stdenv.mkDerivation rec {
+  name = "dmdBuild-${version}";
```
```suggestion
  name = "dmd-${version}";
```

veprbl commented on pkgs/development/tools/build-managers/dub/default.nix:
```diff
@@ -1,109 +1,76 @@
-      ./build.sh
-    '';
+stdenv.mkDerivation rec {
+  name = "dubBuild-${version}";
```
```suggestion
  name = "dub-${version}";
```

veprbl commented on pkgs/development/compilers/ldc/default.nix:
```diff
@@ -18,208 +18,156 @@ let
 
-    enableParallelBuilding = true;
+stdenv.mkDerivation rec {
+  name = "ldcBuild-${version}";
```
```suggestion
  name = "ldc-${version}";
```

ThomasMader commented:
> I've tried to rebuild this on macOS. This time I get a failure in dtools: https://gist.github.com/veprbl/3e0be30bd86142418232f6bf1ff57147

Sorry, forgot to build dtools on macOS. Fixed it by disabling the only test which was using the _-run_ parameter for now.

I also changed the names you suggested.

veprbl commented:
I think you squashed two dub-related commits so the new commit message "dub: Disable network tests because they add impurities" doesn't mention the version bump.

ThomasMader commented:
No dub stayed at the same version because there is no new version yet.
I really disabled only the network tests and the rdmd test because of the _-run_ issue.

veprbl reviewed (approved):
Builds on linux and darwin. `dmd -run` is broken on darwin on current master, so no regression here. This PR fixes dmd build on darwin and improves testing, so this is definitely an improvement.

ThomasMader commented on pkgs/development/compilers/dmd/default.nix:
```diff
@@ -1,199 +1,165 @@
+          --prefix PATH ":" "${targetPackages.stdenv.cc}/bin" \
+          --set-default CC "${targetPackages.stdenv.cc}/bin/cc"
+
+      sed -e "s|outputDirectory|$out|g" ${dmdConfFile} > $out/bin/dmd.conf
```
Fixed

veprbl commented:
@ThomasMader Thank you!
